src/crimson/net/ProtocolV2.h

```diff
@@ -71,6 +76,26 @@ class ProtocolV2 final : public Protocol {
   uint64_t peer_global_seq = 0;
   uint64_t connect_seq = 0;
 
+  seastar::future<> exit_protocol = seastar::now();
+
+  class Timer {
+    double last_dur_ = 0.0;
+    const SocketConnection& conn;
+    std::optional<seastar::abort_source> as;
+   public:
+    Timer(SocketConnection& conn) : conn(conn) {}
+    double last_dur() const { return last_dur_; }
+    seastar::future<> backoff(double dur_s);
+    void cancel() {
+      last_dur_ = 0.0;
+      if (as) {
+        as->request_abort();
+        as = std::nullopt;
+      }
+    }
+  };
+  Timer protocol_timer;
```

Comment:
i am inclined to flatten this class into `ProtocolV2`.

Reply:
The intention of `Timer` is to make it possible to share the same backoff implementation with ProtocolV1.
Do you think it is harder to understand if encapsulate a `Timer` class?

Reply:
oh i see. IIRC, we don't need to focus on the v1 support. will check it with sage tonight.

> Do you think it is harder to understand if encapsulate a `Timer` class?

actually, i am fine either way. i just feel that it's a very thin wrapper around the abortable wait machinery. if we want to reuse it in v1 implementation, probably we can make it more useful by adding the backoff algorithm to it? but could you allow me to check if we want v1 in crimson first?

Reply:
after a second thought, i think we don't need to support lossless connection in v1. as the client's default policy is `Messenger::Policy::lossy_client(CEPH_FEATURE_OSDREPLYMUX)` which is lossy. and by the time of crimson-osd's GA, the in-cluster communication will all be performed using v2 protocol. see 

https://github.com/ceph/ceph/blob/647d80437b64b4be686cec43db24e521223621b0/src/librados/RadosClient.cc#L258

Reply:
Ack, thanks!